src/Microsoft.AspNetCore.Routing/Matchers/CandidateSet.cs

```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Routing.Matchers
+{
+    internal ref struct CandidateSet
+    {
+        public string Path;
+        public ReadOnlySpan<PathSegment> Segments;
+        public ReadOnlySpan<Candidate> Candidates;
+        public ReadOnlySpan<int> CandidateIndices;
+        public ReadOnlySpan<int> CandidateGroups;
+        public Span<RouteValueDictionary> Values;
+
+        public CandidateSet(string path, ReadOnlySpan<PathSegment> segments)
+        {
+            Path = path;
```

Comment:
`Debug.Assert` not null these arguments?

Edit: Now that I've said that in a couple of comments I can see places for `Debug.Assert`s everywhere, which wouldn't be wanted. I guess this is just a prompt to make sure they are where needed.

Reply:
Agreed, I tend to use Debug.Assert liberally where there's complex conditional logic that I can't otherwise refactor. using it everywhere that's possible is just homework.